runtime/elem/Runtime.h

```diff
@@ -332,42 +332,67 @@ namespace elem
             if (!v.isNumber())
+            {
+                ELEM_DBG("[Error] activateRoot - Invalid nodeId format.");
                 return ReturnCode::InvalidInstructionFormat();
+            }
 
-            int32_t nodeId = static_cast<int32_t>((js::Number) v);
+            int32_t nodeId = static_cast<int32_t>((js::Number)v);
             ELEM_DBG("[Native] activateRoot " << nodeIdToHex(nodeId));
 
-            if (nodeTable.find(nodeId) == nodeTable.end())
+            // Using find() method for safe access to nodeTable elements
+            auto it = nodeTable.find(nodeId);
+            if (it == nodeTable.end())
+            {
+                ELEM_DBG("[Error] activateRoot - NodeId not found: " << nodeIdToHex(nodeId));
                 return ReturnCode::NodeNotFound();
+            }
 
-            if (auto ptr = std::dynamic_pointer_cast<RootNode<FloatType>>(nodeTable.at(nodeId))) {
-                ptr->setProperty("active", true);
+            // Attempt to cast the found node to a RootNode type and activate it
+            auto ptr = std::dynamic_pointer_cast<RootNode<FloatType>>(it->second);
+            if (ptr)
+            {
+                ptr->activate(currentRoots.empty() ? FloatType(1) : FloatType(0));
                 active.insert(nodeId);
+                ELEM_DBG("[Success] Activated root: " << nodeIdToHex(nodeId));
+            }
+            else
+            {
+                ELEM_DBG("[Error] Failed to cast to RootNode or activate: " << nodeIdToHex(nodeId));
             }
         }
 
-        // Deactivate any prior roots
-        for (auto const& n : currentRoots) {
-            if (auto ptr = std::dynamic_pointer_cast<RootNode<FloatType>>(nodeTable.at(n))) {
-                // If any current root was not marked active in this event, we deactivate it
-                if (active.count(n) == 0) {
-                    ptr->setProperty("active", false);
-                }
+        // Deactivate any prior roots not included in the incoming active set
+        for (auto const& n : currentRoots)
+        {
+            auto it = nodeTable.find(n);
+            if (it != nodeTable.end())
+            {
+                auto ptr = std::dynamic_pointer_cast<RootNode<FloatType>>(it->second);
+                if (ptr)
+                {
+                    if (active.count(n) == 0)
+                    {
+                        ptr->setProperty("active", false);
+                        ELEM_DBG("[Success] Deactivated root: " << nodeIdToHex(n));
+                    }
 
-                // And if it's still running, we hang onto it
-                if (ptr->stillRunning()) {
-                    active.insert(n);
+                    if (ptr->stillRunning())
+                    {
+                        active.insert(n);
+                    }
                 }
             }
         }
 
-        // Merge
-        currentRoots = active;
+        currentRoots.swap(active); // More efficient than assignment
```

Comment:
I think for the `std::set` it might be a bit more efficient than the assignment because it only deals with internal pointers

Reply:
I think maybe the most clear would be `currentRoots = std::move(active);` ? would do no copy like the call to swap but more clarity. 

though, i really don't think this is a bottleneck at all. 

either way, active is deallocated at the end of the function call. if you really wanted to optimize with swap, active could be a member? 

but imo whatever makes the code clearest here is the best. = seems fine. this function isn't called in a super tight loop or anything

thoughts @cannc4 ?

Reply:
Agreed @ncblair! I see, yea so swap skips the copy whereas otherwise we were using a copy-assignment, got it! I'm ok with either swap or `currentRoots = std::move(active)`